rfcs/2023-rsp-testing-utils.md

```diff
@@ -0,0 +1,113 @@
+
+
+## Detailed Design
+
+As mentioned earlier, there are three proposed parts to this initiative:
+
+1. Documentation of testing gotchas/common issues
+
+There are a number of common mocks that users often have to set up in order to have various components properly render or behave as they do in browser. These include the following:
+
+- Mock of `HTMLElement.prototype`'s `clientWidth`, `clientHeight`, and `scrollHeight`. This is necessary for many of our virtualized collection components like ListBox and TableView. Without these the component often renders with zero height/width and thus zero items. An example of these mocks can be found [here](https://github.com/adobe/react-spectrum/blob/07f673acb2d68144156df3aa0db6e91810b12c67/packages/%40react-spectrum/listbox/test/ListBox.test.js#L71-L76).
+- Mock of `HTMLElement.prototype`'s `getBoundingClientRect`. Often mocked for drag and drop operations, an example can be found [here](https://github.com/adobe/react-spectrum/blob/07f673acb2d68144156df3aa0db6e91810b12c67/packages/%40react-aria/dnd/test/dnd.test.js#L39-L46).
+
+There are other mocks that teams have historically needed, but the above are two of the most commonly brought up. Additionally, the way we've mocked the above in the React Spectrum tests changes it for ALL HTMLElements which maybe too broad for other people's test suites, hence why I'm leaning towards simply documenting this behavior rather than creating a utility function.
+
+In addition to these common mocks, I feel like it would be helpful to compile a list of common issues/error messages people have run into when writing tests using React Spectrum components and add a FAQ section to our existing "Testing" page in the docs. These issues may disappear as we expose more utility functions and classes, but it is a good short term goal that shouldn't take too much time investment.
+
+
+2. Create/expose low level util functions for basic test setup and interactions
+
+Some mocks we use feel like they could be applied universally, such as the mock for `window.screen`'s width when simulating a mobile device. We could expose this as `simulateMobile`/`simulateDesktop` so that users don't have to know the nitty gritty of what breakpoint we use when determining whether to render the mobile version of a React Spectrum component. Another example of a lower level util would be be for triggering a long press event. Instead of needing to know how long it takes before a pointerDown event is considered a long press or that you need to mock support for `PointerEvent` globally, we could provide our internal `triggerLongPress` test util alongside with `installPointerEvent` for mocking `PointerEvent` support. A drag and drop (via mouse and keyboard) interaction utility is another good candidate, as is a render wrapper that wraps arbitrary children in a Provider.
```

Comment:
These would definitely be useful. Not sure whether people will really be simulating mobile/desktop specifically to test RSP components, since they can usually assume that we have already tested that behavior, but no harm in offering it I guess. Long press / pointer event stuff definitely seems useful though.

Reply:
Right, I was mainly thinking along the lines of people writing tests for their app's mobile experience vs desktop experience and thus needing to make our components render in their proper mobile/desktop configuration. I think Quarry had run into this issue previously